File: leetcode/714.买卖股票的最佳时机含手续费/714.买卖股票的最佳时机含手续费.go
package main

import "math"

/*
对比122，增加了手续费。
把手续费从卖出时减去即可
dp[i][0] = max(dp[i-1][0], dp[i-1][1] + prices[i] - fee)
*/
func maxProfit(prices []int, fee int) int {
	n := len(prices)
	dpi0 := 0
	dpi1 := math.MinInt32
	for i := 0; i < n; i++ {
		temp := dpi0
		dpi0 = max(dpi0, dpi1+prices[i]-fee)
		dpi1 = max(dpi1, temp-prices[i])
	}
	return dpi0
}
func max(int1 int, int2 int) int {
	if int1 < int2 {
		return int2
	} else {
		return int1
	}
}
func main() {
	p := []int{1, 3, 2, 8, 4, 9}
	f := 2
	result := maxProfit(p, f)
	print(result)
}

File: leetcode/714.买卖股票的最佳时机含手续费/714.买卖股票的最佳时机含手续费.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        n = len(prices)
        dp_i_0, dp_i_1 = 0, float("-INF")
        for i in range(n):
            temp = dp_i_0
            dp_i_0 = max(dp_i_0, dp_i_1 + prices[i] - fee)
            dp_i_1 = max(dp_i_1, temp - prices[i])
        return dp_i_0


if __name__ == '__main__':
    p = [1, 3, 2, 8, 4, 9]
    f = 2
    result = Solution().maxProfit(p, f)
    print(result)
